plugin-localization/src/main/java/com/mapbox/mapboxsdk/plugins/localization/LocalizationPlugin.java

```diff
@@ -104,20 +122,40 @@ public void setMapLanguage(@NonNull MapLocale mapLocale) {
     for (Source source : mapboxMap.getSources()) {
       if (sourceIsFromMapbox(source)) {
         for (Layer layer : layers) {
-          if (layerHasAdjustableTextField(layer)) {
-            String textField = ((SymbolLayer) layer).getTextField().getValue();
-            if (textField != null
-              && (textField.contains("{name") || textField.contains("{abbr}"))) {
-              textField = textField.replaceAll("[{]((name).*?)[}]",
-                String.format("{%s}", mapLocale.getMapLanguage()));
-              layer.setProperties(textField(textField));
+          if (layer instanceof SymbolLayer) {
+            PropertyValue<?> textFieldProperty = ((SymbolLayer) layer).getTextField();
+            if (textFieldProperty.isExpression()) {
+              convertExpression(mapLocale, layer, textFieldProperty);
+            } else {
+              convertToken(mapLocale, layer, textFieldProperty);
             }
           }
         }
       }
     }
   }
 
+  private void convertToken(@NonNull MapLocale mapLocale, Layer layer, PropertyValue<?> textFieldProperty) {
+    String text = (String) textFieldProperty.getValue();
+    if (text != null && (text.contains(TOKEN_NAME) || text.contains(TOKEN_ABBR))) {
+      layer.setProperties(textField(text.replaceAll(
+        TOKEN_REGEX, String.format(TOKEN_TEMPLATE, mapLocale.getMapLanguage())
+      )));
+    }
+  }
+
+  private void convertExpression(@NonNull MapLocale mapLocale, Layer layer, PropertyValue<?> textFieldProperty) {
+    Expression textFieldExpression = textFieldProperty.getExpression();
+    if (textFieldExpression != null) {
+      String text = textFieldExpression.toString().replaceAll(EXPRESSION_REGEX, mapLocale.getMapLanguage());
+      if (text.startsWith("[\"step") && textFieldExpression.toArray().length % 2 == 0) {
+        // got an invalid step expression from core, we need to add an additional name_x into step
+        text = text.replaceAll(STEP_REGEX, String.format(STEP_TEMPLATE, mapLocale.getMapLanguage()));
```

Comment:
Yep, we’re doing this on iOS/macOS too:

https://github.com/mapbox/mapbox-gl-native/blob/79bf0e8af6bf9ec829a352d56b8e70ccc8f4fa41/platform/darwin/src/NSExpression%2BMGLAdditions.mm#L796-L799